### DISCUSSION TITLE
Backprop function for custom PyTorch models

### DISCUSSION THREAD

**Discussion by ericfeunekes** [explosion/spaCy#7017] [Help: Coding & Implementations] [answered]
This is a two part question about whether I need to define a backprop callback for pytorch models. 

From your documentation [here](https://spacy.io/usage/layers-architectures#frameworks) it looks like I can wrap `nn.Sequential` layers in the pytorch wrapper and use that as a layer in `chain` for example. Is that all I need to do? Or will it need a backprop callback?

If I wanted to create a more complex class-based model ([like this for example](https://github.com/ericfeunekes/Text-Classification-CNN-PyTorch/blob/master/src/model/model.py)), then do I need to create a backprop callback? And will it work if I wrap the class or do I need to structure the model in a different way to work well with the pytorch wrapper?

I've reviewed some of the docs in Thinc and I'm having trouble getting it, unfortunately. Maybe the easiest way to put together a more complex model would be to wrap each pytorch layer and then manipulate them in Thinc? For example, in the model I linked to above I could also see something like this working (though I don't know if it's optimal):
```
from thinc.api import chain, concatenate, with_array, PyTorchWrapper, Relu, Softmax, Model
from typing import List
from thinc.types import Floats2d
from torch import nn

@spacy.registry.architectures("CustomTorchModel.v1")
def create_torch_model(
    nO: int,
    seq_len: int,
    out_size: int,
    kernel_1: int,
    kernel_2: int,
    kernel_3: int,
    kernel_4: int,
    stride: int,
    dropout: float,
) -> Model[List[Floats2d],List[Floats2d]]
	conv_1 = PyTorchWrapper(
				nn.Sequential(
					nn.Conv1d(seq_len, out_size, kernel_1, stride),
					nn.MaxPool1d(kernel_1, stride)
					)
				)
	conv_2 = PyTorchWrapper(
				nn.Sequential(
					nn.Conv1d(seq_len, out_size, kernel_2, stride),
					nn.MaxPool1d(kernel_2, stride)
					)
				)
	conv_3 = PyTorchWrapper(
				nn.Sequential(
					nn.Conv1d(seq_len, out_size, kernel_3, stride),
					nn.MaxPool1d(kernel_3, stride)
					)
				)
	conv_4 = PyTorchWrapper(
				nn.Sequential(
					nn.Conv1d(seq_len, out_size, kernel_4, stride),
					nn.MaxPool1d(kernel_4, stride)
					)
				)

	features = concatenate(conv_1, conv_2, conv_3, conv_4)
	classifier = chain(Relu(dropout=dropout), Softmax())

	model = chain(
				with_array(features),
				classifier
				)
	return model
```

I know there could be some errors hidden in there related to types, etc. But I'm interested in thoughts on whether this will work, conceptually, and whether this is the best way to do it in the context of what I said above. 

I'd want to put this sort of model into a pipeline after a transformer or other tok2vec component and use it as a text classifier. Presumably I need to implement something that will pass in the sequence vectors in a way like that described in the relation extractor video. I haven't gotten to that part yet, still figuring out the wrapper functions.

**Comment by honnibal** [comment, marked as answer]
> From your documentation here it looks like I can wrap nn.Sequential layers in the pytorch wrapper and use that as a layer in chain for example. Is that all I need to do? 

Yep, that should be all you need to do!

Think of it like this. The PyTorch `nn.Sequential` class lets you treat its internals as a black-box. During backprop, it needs to receive a gradient with respect to its outputs, and it'll return you a gradient with respect to its input. All the backprop it does inside itself is its own business. The same is true for pretty much any container in PyTorch. The layers are modular like that. So Thinc doesn't care whether it's an `nn.Sequential` or an `nn.Linear`. It only cares what the input and output types are.

> Or will it need a backprop callback?

If you have a custom PyTorch class that works with custom inputs and outputs, you might need to pass in callbacks to do input and output transformations, so that you can use the pytorch model as a Thinc layer more easily. But you don't always need to do that, and there's generally an alternative solution you could use instead, either changing the PyTorch code or adding a custom Thinc layer.

> For example, in the model I linked to above I could also see something like this working (though I don't know if it's optimal)

That would work fine too. It's potentially less efficient because I expect PyTorch would need to block when we convert the tensor to cupy, but it probably wouldn't be noticeably slower (unsure though).

It just depends how you want to organize your code. If you like using the combinators like `concatenate` or there's some other logic that you find easier to express with the model composition approach, you can definitely do it that way. For that specific example you wrote, I think it would probably be kinder to your future collaborators to just do the concatenation in PyTorch. It's easy to imagine someone reading that code and thinking "wtf is this".
